x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/utils/persistence/SearchAfterDocumentsIterator.java

```diff
@@ -88,6 +97,9 @@ public boolean hasNext() {
         }
 
         SearchResponse searchResponse = doSearch(searchAfterFields());
+        if (trackTotalHits && totalHits.get() == 0) {
+            totalHits.set(searchResponse.getHits().getTotalHits().value);
```

Comment:
Good point. I only track total hits in the first search to avoid possible overhead after the first time. But in an index that is receiving new docs that number may become outdated. Not sure what's the best way to go about this though. I could make an extra search request just for getting the total hits count instead of adding this here. But I'm not sure.

An alternative design would be for the `SearchAfterDocumentsIterator` to allow for a sub-class to add this and then take care of it only in the `TestDocsIterator` where we know we're dealing with a static index.

Open to ideas.

Reply:
You could used `BatchedDocumentsIterator` which uses scroll but testing may take a while and you have to keep the context alive.

Do you expect any changes to the underlying index when this is running? TBH this is only used in the progress meter which are notoriously imprecise. Embrace the imprecision and let that progress meter bounce up and down

Reply:
I ended up adding a comment that explains how total hits are tracked.